clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -2103,6 +2103,42 @@ static void handleCHERIMethodSuffix(Sema &S, Decl *D, const ParsedAttr &Attr) {
   D->addAttr(::new (S.Context) CHERIMethodSuffixAttr(S.Context, Attr, Str));
 }
 
+static void handleCHERIMMIODevice(Sema &S, Decl *D, const ParsedAttr &Attr,
+                                  Sema::DeclAttributeLocation DAL) {
+  StringRef DeviceName;
+  SourceLocation DeviceNameLiteralLoc;
+  if (!S.checkStringLiteralArgumentAttr(Attr, 0, DeviceName,
+                                        &DeviceNameLiteralLoc))
+    return;
+  StringRef EncodedPermissions;
+  SourceLocation EncodedPermissionsLiteralLoc;
+
+  if (!S.checkStringLiteralArgumentAttr(Attr, 1, EncodedPermissions,
+                                        &EncodedPermissionsLiteralLoc))
+    return;
+
+  D->addAttr(::new (S.Context) CHERIMMIODeviceAttr(S.Context, Attr, DeviceName,
+                                                   EncodedPermissions));
+}
+
+static void handleCHERISharedObject(Sema &S, Decl *D, const ParsedAttr &Attr,
+                                    Sema::DeclAttributeLocation DAL) {
+  StringRef ObjectName;
+  SourceLocation ObjectNameLiteralLoc;
+  if (!S.checkStringLiteralArgumentAttr(Attr, 0, ObjectName,
+                                        &ObjectNameLiteralLoc))
+    return;
+  StringRef EncodedPermissions;
+  SourceLocation EncodedPermissionsLiteralLoc;
```

Comment:
Same comment about validating permissions.

Reply:
https://github.com/CHERIoT-Platform/llvm-project/pull/147/commits/4ffcc2cb376e395259065c8636b29833bfce7139 adds the validation.